src/hazard_manager/mod.rs

```diff
@@ -0,0 +1,439 @@
+            .or_else(|| Self::u32_as_and(val))
+            .or_else(|| Self::u32_as_cbz(val))
+            .or_else(|| Self::u32_as_cmp(val))
+            .or_else(|| Self::u32_as_bcs(val))
+            .or_else(|| Self::u32_as_ldrsw(val))
+            .unwrap_or(Self::Unk(val))
+    }
+}
+
+enum HazardState {
+    Begin,
+    Adrp1,
+    Add2,
+    Ldur3,
+    Ldrb4,
+    Ldr5
+}
+
+enum HookState {
+    Begin,
+    Adrp1,
+    Ldrsw2
+}
+
+use HookState::*;
+use HazardState::*;
+use Instruction::*;
+
+fn adrp_get_imm(instr: u32) -> u32 {
+    let immhi = (instr >> 5) & 0x7FFFF;
+    let immlo = (instr >> 29) & 0x3;
+    return (immhi << 14) + (immlo << 12);
+}
+
+fn add_get_imm(instr: u32) -> u32 {
+    return (instr >> 10) & 0xFFF;
+}
+
+fn get_hazard_flag_address() -> usize {
+    let mut state = HazardState::Begin;
+    let mut flag_pos = 0;
+    for (pos, instr) in TextIter::new() {
+        state = match (state, instr) {
+            (HazardState::Begin, Adrp { .. }) => {
+                flag_pos = pos;
+                HazardState::Adrp1
+            },
+            (HazardState::Adrp1, Add { .. }) => Add2,
+            (Add2, Ldur { .. }) => Ldur3,
+            (Ldur3, Ldrb { .. }) => Ldrb4,
+            (Ldrb4, Ldr { .. }) => Ldr5,
+            (Ldr5, Sub { .. }) => {
+                break;
+            },
+            _ => {
+                flag_pos = 0;
+                HazardState::Begin
+            }
+        }
+    }
+    if flag_pos == 0 {
+        return 0x0;
+    }
+    let program_counter = flag_pos & !0xFFF; // Need program counter to mimic ADRP
+    let adrp = unsafe { adrp_get_imm(*(flag_pos as *mut u32)) as usize };
+    let add = unsafe { add_get_imm(*((flag_pos + 4) as *mut u32)) as usize };
+    return program_counter + adrp + add + 0x9;
+}
+
+fn get_hazard_hook_address() -> usize {
+    let mut state = HookState::Begin;
+    let mut flag_pos = 0;
+    for (pos, instr) in TextIter::new() {
+        state = match (state, instr) {
+            (HookState::Begin, Adrp { .. }) => HookState::Adrp1,
+            (HookState::Adrp1, Ldrsw { .. }) => Ldrsw2,
+            (Ldrsw2, Cbz { .. }) => {
+                flag_pos = pos;
+                break;
+            }
+            _ => {
+                flag_pos = 0;
+                HookState::Begin
+            }
+        }
+    }
+    return flag_pos;
+}
+
+// 8.1.0 Defaults
+static mut HAZARD_FLAG_ADDRESS: *mut u8 = 0x4ebbf95 as *mut u8;
+static mut LOAD_ADDRESS: usize = 0x214bde8;
+
+#[hook(offset = LOAD_ADDRESS, inline)]
+fn hazard_intercept(ctx: &skyline::hooks::InlineCtx) {
+    unsafe {
+        if is_training_mode() {
+            mod_handle_hazards();
+        }
+    }
+}
+
+fn mod_handle_hazards() {
+    unsafe {
+        if MENU.stage_hazards == StageHazards::On {
+            *HAZARD_FLAG_ADDRESS = 0x1;
+        }
+        else {
+            *HAZARD_FLAG_ADDRESS = 0x0;
+        }
+    }
+}
+
+fn display_hazards_error(is_flag_addr: bool) {
```

Comment:
This is probably better as `validate_hazards_addrs` function that takes no args, and does the ADDR checking.